src/details/ArborX_MinimumSpanningTree.hpp

```diff
@@ -0,0 +1,504 @@
+/****************************************************************************
+ * Copyright (c) 2017-2021 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_MINIMUM_SPANNING_TREE_HPP
+#define ARBORX_MINIMUM_SPANNING_TREE_HPP
+
+#include <ArborX_AccessTraits.hpp>
+#include <ArborX_DetailsKokkosExtArithmeticTraits.hpp>
+#include <ArborX_DetailsKokkosExtMinMaxOperations.hpp>
+#include <ArborX_DetailsMutualReachabilityDistance.hpp>
+#include <ArborX_DetailsTreeNodeLabeling.hpp>
+#include <ArborX_DetailsUtils.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX
+{
+namespace Details
+{
+
+struct WeightedEdge
+{
+  int source;
+  int target;
+  float weight;
+
+private:
+  // performs lexicographical comparison by comparing first the weights and then
+  // the unordered pair of vertices
+  friend KOKKOS_FUNCTION constexpr bool operator<(WeightedEdge const &lhs,
+                                                  WeightedEdge const &rhs)
+  {
+    if (lhs.weight < rhs.weight)
+    {
+      return true;
+    }
+    if (lhs.weight > rhs.weight)
+    {
+      return false;
+    }
+    using KokkosExt::min;
+    auto const lhs_min = min(lhs.source, lhs.target);
+    auto const rhs_min = min(rhs.source, rhs.target);
+    if (lhs_min < rhs_min)
+    {
+      return true;
+    }
+    if (lhs_min > rhs_min)
+    {
+      return false;
+    }
+    using KokkosExt::max;
+    auto const lhs_max = max(lhs.source, lhs.target);
+    auto const rhs_max = max(rhs.source, rhs.target);
+    return (lhs_max < rhs_max);
+  }
+};
+
+template <class BVH, class Labels, class Edges, class Metric, class Radii>
+struct FindComponentNearestNeighbors
+{
+  BVH bvh_;
+  Labels labels_;
+  Edges edges_;
+  Metric mutual_reachability_distance_;
+  Radii radii_;
+
+  template <class ExecutionSpace>
+  FindComponentNearestNeighbors(ExecutionSpace const &space, BVH const &bvh,
+                                Labels const &labels, Edges const &edges,
+                                Metric const &mutual_reachability_distance,
```

Comment:
Rename `mutual_reachability_distance` to something more generic, as it could also be euclidean distance.

Reply:
I thought about it and, we can't call it "distance", and I thought that was an acceptable name.  What alternative would you like to suggest?